### ISSUE TITLE
Support field retrieval joins

### ISSUE THREAD

**Issue by dnhatn** [elastic/elasticsearch#76133]
Today, the [enrich processor ](https://www.elastic.co/guide/en/elasticsearch/reference/current/enrich-processor.html) supports enriching data seamlessly at indexing time. But users must reindex whenever the enrich index changes. This issue proposes to support enriching field retrieval at query time as an alternative. The proposal only supports field retrieval joins; searches and aggregation aren't covered. Like [runtime fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/runtime.html), field retrieval joins also reduce disk usage and indexing time because enriched fields are not looked up and indexed. However, field retrieval joins are not suitable if they are retrieved often or searched or aggregated.

Below is an example of the proposed API. It supports look up by the` _id` of the target index.

```js
PUT /users/_doc/mardy.brown?refresh=wait_for
{
  "email": "mardy.brown@asciidocsmith.com",
  "first_name": "Mardy",
  "last_name": "Brown",
  "city": "New Orleans",
  "county": "Orleans",
  "state": "LA",
  "zip": 70116,
  "web": "mardy.asciidocsmith.com",
  "joined_date": "2021-01-02"
}
```

```js
POST /blogs/_doc?refresh=wait_for
{
  "title": "Elasticsearch is fast",
  "author": "mardy.brown"
}
```

```js
POST /blogs/_search
{
  "query": {
    "match": {
      "title": "fast"
    }
  },
  "fields": [
    "title",
    {
      "field": "author.first_name",
      "lookup_index": "users",
      "match_field": "author",
      "lookup_path": "first_name" 
    },
    {
      "field": "author.last_name", // optional, if not specified then it's `{{lookup_index}}.{{lookup_path}}`
      "lookup_index": "users",
      "match_field": "author",
      "lookup_path": "last_name"
    },
    {
      "field": "author.joined_date",
      "lookup_index": "users",
      "match_field": "author",
      "format": "yyyy-MM-dd",
      "lookup_path": "joined_date"
    }
  ]
}
```

Should we also support queries in the lookup? I can be useful, but I don't find a good use case.

```js
POST /blogs/_search
{
  "query": {
    "match": {
      "title": "cool"
    }
  },
  "fields": [
    "title",
    {
      "field": "first_name",
      "lookup_index": "users",
      "query" : {
          "match": {
              "email": "{{email}}"
          }
      }
    }
  ]
}
```

Should we also support field retrieval from the aggregated result?

```js
GET restaurant/_search
{
  "query": {
    "match": {
      "menu": "sushi"
    }
  },
  "fields": [
    "name",
    {
      "field": "rating",
      "lookup_index": "reviews",
      "query": {
        "term": {
          "_id": "{{restaurant_id}}"
        }
      },
      "aggs": {
        "avg_rating":{"avg": { "field": "rating" }}
      },
      "aggs_path": "aggregations.avg_rating.value"
    }
  ]
}
```

A field retrieval joins can be added to the mapping of an index.

Feedback on the API and these questions are welcomed.

**Comment by giladgal** [comment]
@AlonaNadler @oatkiller to verify this is useful for your use cases. 
For the ad-hoc enrichment use case in Kibana, I think the workflow will be to load the data, create an index with it, and then define the join and enrichment. I believe this will be a set of operations that will each be invoked separately by the user from Kibana.

**Comment by ywelsch** [comment]
A scenario I was wondering about was multi-level field retrieval joins (e.g. blog -> author/users -> organization/companies). With the structure that Nhat proposed in https://github.com/elastic/elasticsearch/pull/76905 the structure can simply be nested and define a computation/lookup tree. With the "runtime_mappings" approach, how would one define this? "match_field" possibly referring to a runtime (lookup) field?

**Comment by dnhatn** [comment]
We have implemented this as lookup runtime fields in #82385.
